edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsScraper.cs

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net;
+    using System.Net.Http;
+    using System.Text.RegularExpressions;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+
+    public class MetricsScraper : IMetricsScraper, IDisposable
+    {
+        const string UrlPattern = @"[^/:]+://(?<host>[^/:]+)(:[^:]+)?$";
+        static readonly Regex UrlRegex = new Regex(UrlPattern, RegexOptions.Compiled);
+        readonly HttpClient httpClient;
+        readonly Lazy<IDictionary<string, string>> endpoints;
+        readonly ISystemTime systemTime;
+        static readonly ILogger Log = Logger.Factory.CreateLogger<MetricsScraper>();
+
+        public MetricsScraper(IList<string> endpoints, ISystemTime systemTime = null)
+        {
+            this.httpClient = new HttpClient();
+            this.endpoints = new Lazy<IDictionary<string, string>>(() => endpoints.ToDictionary(e => e, this.GetUriWithIpAddress));
+            this.systemTime = systemTime ?? SystemTime.Instance;
+        }
+
+        public Task<IEnumerable<Metric>> ScrapeAsync(CancellationToken cancellationToken)
+        {
+            return this.endpoints.Value.SelectManyAsync(async endpoint =>
+            {
+                Log.LogInformation($"Scraping endpoint {endpoint.Key}");
+                string metricsData = await this.ScrapeEndpoint(endpoint.Value, cancellationToken);
```

Comment:
I think await here will stop and wait for the result come back.  SelectManyAsync may not be able to return all tasks immediately.

Reply:
It will wait until all endpoints have returned.